src/DataGrid.tsx

```diff
@@ -958,25 +932,44 @@ function DataGrid<R, SR, K extends Key>(
     );
   }
 
-  function getRowViewportColumns(rowIdx: number) {
-    const selectedColumn = columns[selectedPosition.idx];
+  function addOutOfBoundColumnToRowViewportColumns(
+    columnToAddIdx: number,
+    rowViewportColumns: readonly CalculatedColumn<R, SR>[]
+  ) {
+    const column = columns[columnToAddIdx];
     if (
       // idx can be -1 if grouping is enabled
       // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-      selectedColumn !== undefined &&
-      selectedPosition.rowIdx === rowIdx &&
-      !viewportColumns.includes(selectedColumn)
+      column !== undefined &&
+      !rowViewportColumns.includes(column)
     ) {
       // Add the selected column to viewport columns if the cell is not within the viewport
-      return selectedPosition.idx > colOverscanEndIdx
-        ? [...viewportColumns, selectedColumn]
+      return columnToAddIdx > colOverscanEndIdx
+        ? [...rowViewportColumns, column]
         : [
-            ...viewportColumns.slice(0, lastFrozenColumnIndex + 1),
-            selectedColumn,
-            ...viewportColumns.slice(lastFrozenColumnIndex + 1)
+            ...rowViewportColumns.slice(0, lastFrozenColumnIndex + 1),
+            column,
+            ...rowViewportColumns.slice(lastFrozenColumnIndex + 1)
           ];
     }
-    return viewportColumns;
+    return rowViewportColumns;
+  }
+
+  function getRowViewportColumns(rowIdx: number) {
+    let rowViewportColumns = viewportColumns;
+    if (rowIdx === selectedPosition.rowIdx) {
+      rowViewportColumns = addOutOfBoundColumnToRowViewportColumns(
+        selectedPosition.idx,
+        viewportColumns
+      );
+    }
+    if (rowIdx === -1 && scrollToColumnIdx !== undefined) {
```

Comment:
Same logic as the selected cell. We need to add the column if it is not in the viewport